From: Rafael J. Wysocki <rjw@suse.de>
Subject: x86_64 / PCI: Insert IO-APIC resources before unassigned resources
References: bnc#531260
Patch-upstream: Yes

Based on upstream commit 857fdc53a0a90c3ba7fcf5b1fb4c7a62ae03cf82
(x86/pci: insert ioapic resource before assigning unassigned
resources) from Yinghai Lu <yinghai@kernel.org>.

It turns out that:
(1) AMD-based systems have two HT chains.
(2) BIOS doesn't allocate resources for BAR 6 of devices under 8132
    etc.
(3) Multi-peer-root patch will try to split root resources to peer
    root resources according to PCI conf of NB
(4) PCI core assigns unassigned resources, but they overlap with BARs
    that are used by IO-APIC addr of io4 and 8132.

This happens because at that point the IO-APIC resources have not
been inserted yet.  The solution is to insert IO-APIC resources into
the tree a bit earlier.

Signed-off-by: Rafael J. Wysocki <rjw@suse.de>
---
 arch/x86/kernel/io_apic_64.c |   10 ++--------
 arch/x86/pci/i386.c          |    7 +++++++
 include/asm-x86/io_apic.h    |    4 ++++
 3 files changed, 13 insertions(+), 8 deletions(-)

Index: linux-2.6.27-SLE11_BRANCH/include/asm-x86/io_apic.h
===================================================================
--- linux-2.6.27-SLE11_BRANCH.orig/include/asm-x86/io_apic.h
+++ linux-2.6.27-SLE11_BRANCH/include/asm-x86/io_apic.h
@@ -211,12 +211,16 @@ extern void ioapic_init_mappings(void);
 extern int save_mask_IO_APIC_setup(void);
 extern void restore_IO_APIC_setup(void);
 extern void reinit_intr_remapped_IO_APIC(int);
+extern void ioapic_insert_resources(void);
+#else
+static inline void ioapic_insert_resources(void) { }
 #endif
 
 #else  /* !CONFIG_X86_IO_APIC */
 #define io_apic_assign_pci_irqs 0
 static const int timer_through_8259 = 0;
 static inline void ioapic_init_mappings(void) { }
+static inline void ioapic_insert_resources(void) { }
 #endif
 
 #endif
Index: linux-2.6.27-SLE11_BRANCH/arch/x86/kernel/io_apic_64.c
===================================================================
--- linux-2.6.27-SLE11_BRANCH.orig/arch/x86/kernel/io_apic_64.c
+++ linux-2.6.27-SLE11_BRANCH/arch/x86/kernel/io_apic_64.c
@@ -3030,7 +3030,7 @@ void __init ioapic_init_mappings(void)
 	}
 }
 
-static int __init ioapic_insert_resources(void)
+void __init ioapic_insert_resources(void)
 {
 	int i;
 	struct resource *r = ioapic_resources;
@@ -3038,18 +3038,12 @@ static int __init ioapic_insert_resource
 	if (!r) {
 		printk(KERN_ERR
 		       "IO APIC resources could be not be allocated.\n");
-		return -1;
+		return;
 	}
 
 	for (i = 0; i < nr_ioapics; i++) {
 		insert_resource(&iomem_resource, r);
 		r++;
 	}
-
-	return 0;
 }
 
-/* Insert the IO APIC resources after PCI initialization has occured to handle
- * IO APICS that are mapped in on a BAR in PCI space. */
-late_initcall(ioapic_insert_resources);
-
Index: linux-2.6.27-SLE11_BRANCH/arch/x86/pci/i386.c
===================================================================
--- linux-2.6.27-SLE11_BRANCH.orig/arch/x86/pci/i386.c
+++ linux-2.6.27-SLE11_BRANCH/arch/x86/pci/i386.c
@@ -33,6 +33,7 @@
 #include <linux/bootmem.h>
 
 #include <asm/pat.h>
+#include <asm/io_apic.h>
 
 #include "pci.h"
 
@@ -227,6 +228,12 @@ void __init pcibios_resource_survey(void
 	pcibios_allocate_bus_resources(&pci_root_buses);
 	pcibios_allocate_resources(0);
 	pcibios_allocate_resources(1);
+	/*
+	 * Insert the IO APIC resources after PCI initialization has
+	 * occured to handle IO APICS that are mapped in on a BAR in
+	 * PCI space, but before trying to assign unassigned pci res.
+	 */
+	ioapic_insert_resources();
 }
 
 /**
